minetest/minetest
# Various visual effects

This PR adds various effects intended to improve the visual experience. This includes changes to the shader code, as well as the engine source code, Lua API and settings.

<details>
<summary>List of Features and Changes</summary>

- Tinted sunlight during dusk and dawn
- Options for warmer artificial light colors
- "soft" clouds look
- Slight overall changes to the post processing shaders
- Crude water reflections (sky and sun) and waves
- Color grading
- Vignette
- Translucent foliage
- Bump maps
- Node "reflections" (just reflections of sunlight)
- Adjusted fog color (more saturated where the fog is lighter)

### Lua API additions
- Cloud parameters: `shadow` field controls the color of the cloud's base
- Lighting: `shadow_tint` controls shadow color
    - default would be black `{ r = 0, g = 0, b = 0}`
    - ambient light from the sky could be emulated by a blue color like `{ r = 0, g = 50, b = 180 }`

### Settings
| Name | Technical Name | Location | Type | Default |
|---|---|---|---|---|
| Tinted sunlight | tinted_sunlight | Graphics>Graphics Effects | bool | false |
| Artificial light color | artificial_light | Graphics>Graphics Effects | enum | "default" |
| Soft clouds | soft_clouds | Graphics>Clouds | bool | false |
| Liquid reflections | enable_water_reflections | Shaders>Waving Nodes | bool | false |
| Color grading | enable_color_grading | Shaders>Post Processing | bool | false |
| Vignette | enable_vignette | Shaders>Post Processing | bool | false |
| Translucent Foliage | enable_translucent_foliage | Shaders>Other Effects | bool | false |
| Bump maps | enable_bump_maps | Shaders>Other Effects | bool | false |
| Node reflections | enable_node_reflections | Shaders>Other Effects | bool | false |
</details>

## To do

This PR is Ready for Review.

## How to test

For most features, simply change the game settings. Of course, the game you use should support dynamic shadows and volumetrics.
The added Lua properties can be used like so:
```Lua
player:set_clouds({shadow = { r = 120, g = 160, b = 240 }})
player:set_lighting({shadow_tint = { r = 0, g = 50, b = 180 }})
```
